lucetc/src/compiler.rs

```diff
@@ -47,6 +47,91 @@ impl OptLevel {
     }
 }
 
+pub struct CompilerBuilder {
+    pub target: Triple,
+    opt_level: OptLevel,
+    cpu_features: CpuFeatures,
+    heap_settings: HeapSettings,
+    count_instructions: bool,
+    canonicalize_nans: bool,
+    validator: Option<Validator>,
+}
+
+impl CompilerBuilder {
+    pub fn new() -> Self {
+        Self {
+            target: Triple::host(),
+            opt_level: OptLevel::default(),
+            cpu_features: CpuFeatures::default(),
+            heap_settings: HeapSettings::default(),
+            count_instructions: false,
+            canonicalize_nans: false,
+            validator: None,
+        }
+    }
+
+    pub fn target(&mut self, target: Triple) -> &mut Self {
```

Comment:
As you mentioned above, this could follow the `LucetcOpts` conventions :slightly_smiling_face: 

Instead of having this borrow the struct, we can have it consume `self`, and return an owned type, via the signature below. This just means passing ownership of the object along, instead of passing along the ability to mutate the object. Call-sites shouldn't change _too_ much. :crossed_fingers: 

```rust
pub fn with_target(self, target: Triple) -> Self
```

Reply:
if I try this signature, then within `LucetOpts` impl, it complains that `CompilerBuilder` needs to implement `Copy`.  Lots of bits in there don't implement `Copy` (`Triple`, `CpuFeatures`, `HeapSettings`) so I'm wondering if I'm missing something about how to do this.  I see how the impl of `LucetOpts` uses the following signature

 ```rust
fn with_target(mut self, target: Triple) -> Self
```
but not sure why it doesn't need the `Copy`. Can dig more.

Reply:
ok, i think i have this how you're asking @katie-martin-fastly.  Note there is still the `_mut` versions as below.  as before, there are a lot more places to clean up if this looks good.